### DISCUSSION TITLE
Warn when a record includes a member that can't use value equality

### DISCUSSION THREAD

**Discussion by rlundy** [dotnet/csharplang#7339] [Language Ideas]
The C# documentation on records generally states that they use value equality.  See here, for instance:

https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/record#value-equality

> For types with the record modifier (record class, record struct, and readonly record struct), two objects are equal if they are of the same type and store the same values.

This isn't quite true, because if a record contains a collection property, for instance, the collection items are not compared for equality.  It appears that the collection objects are compared by reference.

Using xUnit.net, here's a quick sample of the problem:

```
public class ComparingSomeRecords
{
    record Stuff
    {
        public List<string> SomeText = new();
    }

    [Fact]
    void CanRecordsWithCollectionsCompareCorrectly()
    {
        var stuff1 = new Stuff();
        var stuff2 = new Stuff();

        Assert.True(stuff1.Equals(stuff2));
    }
}
```

The test fails because the List<string> instances aren't considered equal.

It would be helpful if there were a compiler warning when a record includes a member that doesn't use value equality, as above.

**Comment by jcouv** [comment]
Moved issue to csharplang since the current implementation is by-design.
As far as the proposal for a new warning, I think: 1. this is best left to an analyzer, 2. it's not completely clear what the rules would be.

**Comment by huoyaoyuan** [comment]
The definition of 
> store the same values

should be reference equality for non-record types. The documentation should clarify this.

**Comment by CyrusNajmabadi** [comment]
> should be reference equality 

It's not reference equality.  It's whatever equality semantics that particular type has. We defer to the individual members to decide.

**Comment by rlundy** [comment]
I guess "compiler warning" was not the right thing to say.  An analyzer warning would indeed make more sense.  Or just a documentation improvement to make much more clear the limits of so-called "value equality" in a record.

**Reply by RikkiGibson** [reply]
fwiw, I absolutely feel that records containing collection types are both a pit of failure and quite common. See discussion in https://github.com/dotnet/runtime/issues/77183#issuecomment-1283195266.

While I'd like to investigate further at some point, I don't currently know what if anything can be done on the language level to address the situation.

So, in any case, I would love to see movement on an analyzer here. Presumably, one which could identify some set of types whose equality semantics are unlikely to be what the user wants in a record. Some careful thought would be needed to arrive at the right design for that.
